src/data/finset/noncomm_prod.lean

```diff
@@ -0,0 +1,158 @@
+      { intros hl hl' _,
+        rw heq_iff_eq,
+        refine h.prod_eq' _,
+        rw list.pairwise.imp_mem,
+        refine list.pairwise_of_forall _,
+        intros x y hx hy,
+        by_cases hxy : x = y,
+        { subst hxy },
+        { exact hl x hx y hy hxy } }
+    end)
+
+@[simp, to_additive] lemma noncomm_prod_coe (l : list α) (comm) :
+  noncomm_prod (l : multiset α) comm = l.prod := rfl
+
+@[to_additive]
+lemma noncomm_prod_congr_list_prod (s : multiset α) (l : list α) (comm) (hl : s = l) :
+  s.noncomm_prod comm = l.prod :=
+begin
+  induction s using quotient.induction_on,
+  simp only [quot_mk_to_coe, coe_eq_coe] at hl,
+  simp only [quot_mk_to_coe, noncomm_prod_coe],
+  refine hl.prod_eq' _,
+  rw list.pairwise.imp_mem,
+  refine list.pairwise_of_forall _,
+  intros x y hx hy,
+  by_cases hxy : x = y,
+  { subst hxy },
+  { exact comm _ hx _ hy hxy }
+end
+
+@[simp, to_additive] lemma noncomm_prod_empty (h) :
+  noncomm_prod (0 : multiset α) h = 1 := rfl
+
+@[simp, to_additive] lemma noncomm_prod_cons (s : multiset α) (a : α) (comm) :
+  noncomm_prod (a ::ₘ s) comm = a * noncomm_prod s
+    (λ x hx y hy, comm _ (mem_cons_of_mem hx) _ (mem_cons_of_mem hy)) :=
+begin
+  induction s using quotient.induction_on,
+  simp
+end
+
+@[to_additive] lemma noncomm_prod_cons' (s : multiset α) (a : α) (comm) :
+  noncomm_prod (a ::ₘ s) comm = noncomm_prod s
+    (λ x hx y hy, comm _ (mem_cons_of_mem hx) _ (mem_cons_of_mem hy)) * a :=
+begin
+  induction s using quotient.induction_on with s,
+  simp only [quot_mk_to_coe, cons_coe, noncomm_prod_coe, list.prod_cons],
+  induction s with hd tl IH,
+  { simp },
+  { rw [list.prod_cons, mul_assoc, ←IH, ←mul_assoc, ←mul_assoc],
+    { congr' 1,
+      by_cases ha : a = hd,
+      { subst ha },
+      { refine comm _ _ _ _ ha,
+        { simp },
+        { simp } } },
+    { intros x hx y hy,
+      simp only [quot_mk_to_coe, list.mem_cons_iff, mem_coe, cons_coe] at hx hy,
+      apply comm,
+      { cases hx;
+        simp [hx] },
+      { cases hy;
+        simp [hy] } } }
+end
+
+end multiset
+
+namespace finset
+
+/-- Product of a `s : finset α` with `[monoid α`], given a proof that `*` commutes
+on all elements `x ∈ s`. -/
+@[to_additive "Sum of a `s : finset α` with `[add_monoid α`], given a proof that `*` commutes
+on all elements `x ∈ s`."]
+def noncomm_prod (s : finset α) (comm) : α :=
+s.1.noncomm_prod comm
+
+@[simp, to_additive] lemma noncomm_prod_to_finset [decidable_eq α] (l : list α)
+  (comm) (hl : l.nodup) :
+  noncomm_prod l.to_finset comm = l.prod :=
+begin
+  rw ←list.erase_dup_eq_self at hl,
+  simp [noncomm_prod, hl]
+end
+
+@[to_additive]
+lemma noncomm_prod_congr_list_prod [decidable_eq α] (s : finset α) (l : list α)
+  (comm) (hl : s = l.to_finset) (hn : l.nodup) :
+  s.noncomm_prod comm = l.prod :=
+begin
+  rw noncomm_prod,
+  rw ←list.erase_dup_eq_self at hn,
+  refine multiset.noncomm_prod_congr_list_prod _ _ _ _,
+  simp [hl, hn]
+end
+
```

Comment:
```suggestion
@[to_additive]
lemma noncomm_prod_congr_list_prod [decidable_eq α] (s : finset α) (l : list α)
  (comm) (hl : s = l.to_finset) (hn : l.nodup) :
  s.noncomm_prod comm = l.prod :=
begin
  subst hl,
  exact noncomm_prod_to_finset _ _ hn,
end

```
Do you have an example of where this lemma is more convenient to use than `noncomm_prod_to_finset`?